### ISSUE TITLE
Intrepid2: Significant Performance Regression in CUDA Integration

### ISSUE THREAD

**Issue by CamelliaDPG** [trilinos/Trilinos#10477]
## Bug Report
@trilinos/intrepid2

### Description
Sometime between Summer 2021 and now, performance of "standard" Intrepid2 integration has suffered dramatically.  Whereas previously we were estimating about 200 GFlops/sec for our core integration kernels for e.g. p=5, now we see about 7 GFlops/sec.  (I'm running this on Weaver.)

I've distilled this down a little, but have not yet done e.g. a git bisect to work out precisely when the slowdown happened.

I have two forked branches that demonstrate the issue:
1. [slowdown-baseline-Trilinos-summer-2021-try-2](https://github.com/CamelliaDPG/Trilinos/tree/slowdown-baseline-Trilinos-summer-2021-try-2) 
2. [slowdown-regression-demo-current-Trilinos](https://github.com/CamelliaDPG/Trilinos/tree/slowdown-regression-demo-current-Trilinos)

The first will show the previous performance; the second will show current performance.

Here's the previous performance:
```
[nvrober@weaver4 weaver-fork-Trilinos-intrepid2-release-with-uvm]$ ./packages/intrepid2/unit-test/performance/StandardIntegration/Intrepid2_unit-test_performance_StandardIntegration_StandardIntegrationPerformance.exe --polyOrder=5 --mode=BestCuda
Teuchos::GlobalMPISession::GlobalMPISession(): started serial run


Running with polyOrder = 5, meshWidth = 16
Standard, workset size:                        8
Time (core integration)                 1.18e+00 seconds (57.87%).
flop estimate (core):                   2.50e+11
estimated throughput (core):            2.12e+02 GFlops
Time (Jacobians)                        7.93e-01 seconds (38.81%).
flop estimate (Jacobians):              1.90e+08
estimated throughput (Jac.):            2.40e-01 GFlops
Time (initial setup)                    5.34e-02 seconds (2.62%).
Time (other)                            1.43e-02 seconds (0.70%).
Time (total):                           2.04e+00 seconds.
flop estimate (total):                  2.51e+11
estimated throughput (total):           1.23e+02 GFlops
```

Here's the current:
```
[nvrober@weaver4 weaver-fork-Trilinos-intrepid2-release-with-uvm]$ ./packages/intrepid2/unit-test/performance/StandardIntegration/Intrepid2_unit-test_performance_StandardIntegration_StandardIntegrationPerformance.exe --polyOrder=5 --mode=BestCuda
Teuchos::GlobalMPISession::GlobalMPISession(): started serial run


Running with polyOrder = 5, meshWidth = 16
Standard, workset size:                        8
Time (core integration)                 3.85e+01 seconds (95.71%).
flop estimate (core):                   2.50e+11
estimated throughput (core):            6.50e+00 GFlops
Time (Jacobians)                        1.56e+00 seconds (3.89%).
flop estimate (Jacobians):              1.90e+08
estimated throughput (Jac.):            1.22e-01 GFlops
Time (initial setup)                    1.45e-01 seconds (0.36%).
Time (other)                            1.96e-02 seconds (0.05%).
Time (total):                           4.03e+01 seconds.
flop estimate (total):                  2.51e+11
estimated throughput (total):           6.22e+00 GFlops
```

The branch [slowdown-baseline-Trilinos-summer-2021-try-2](https://github.com/CamelliaDPG/Trilinos/tree/slowdown-baseline-Trilinos-summer-2021-try-2) was created by checking out `df50af1e7c27415ebd83da733fd7f18021a46841`, and then adding the new `StandardIntegrationPerformance` executable.

The branch [slowdown-regression-demo-current-Trilinos](https://github.com/CamelliaDPG/Trilinos/tree/slowdown-regression-demo-current-Trilinos) was created by branching from Trilinos develop, and adding the same `StandardIntegrationPerformance` executable.

### Steps to Reproduce
1. Do a CUDA build of `StandardIntegrationPerformance` from the branch [slowdown-baseline-Trilinos-summer-2021-try-2](https://github.com/CamelliaDPG/Trilinos/tree/slowdown-baseline-Trilinos-summer-2021-try-2).
1. Run `./packages/intrepid2/unit-test/performance/StandardIntegration/Intrepid2_unit-test_performance_StandardIntegration_StandardIntegrationPerformance.exe --polyOrder=5 --mode=BestCuda`
1. Do a CUDA build of `StandardIntegrationPerformance` from the branch [slowdown-regression-demo-current-Trilinos](https://github.com/CamelliaDPG/Trilinos/tree/slowdown-regression-demo-current-Trilinos)
1. Run `./packages/intrepid2/unit-test/performance/StandardIntegration/Intrepid2_unit-test_performance_StandardIntegration_StandardIntegrationPerformance.exe --polyOrder=5 --mode=BestCuda`

**Comment by GrahamBenHarper** [comment]
@CamelliaDPG I just realized the slowdown branch uses Kokkos 3.4, and there was supposedly a large performance issue fixed in Kokkos 3.6 (see slide 21/58 in https://github.com/kokkos/kokkos-tutorials/blob/main/Other/ReleaseBriefings/release-36.pdf).

It might be worthwhile to test things out with the updated Kokkos 3.6 on develop at some point.

**Comment by CamelliaDPG** [comment]
@GrahamBenHarper, that sounds like a great lead.  I'll try with Kokkos 3.6 sometime soon, and report results back here.
